### PR TITLE
Global coarsening: compress weights

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

**Comment by peterrum** [comment]
@kronbichler I am tempted to get rid of the short-cut path, since it makes the application of hanging-node constraints and the application of weights annoyingly complicated. What do you think?

https://github.com/dealii/dealii/blob/cfa11cfe8cbc78f9a3b15bdb435fd2297d4b871d/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h#L2526-L2571

**Comment by kronbichler** [comment]
> I am tempted to get rid of the short-cut path [...]

I think I agree. Just so we do not miss anything, can you summarize the operations we need to go through in that case? I guess it is a single sum-factorization interpolation (at least if we take the right path), i.e., approximately `dim * (n_points_fine)^dim n_points_coarse`?

**Comment by peterrum** [comment]
> I think I agree. Just so we do not miss anything, can you summarize the operations we need to go through in that case? I guess it is a single sum-factorization interpolation (at least if we take the right path), i.e., approximately dim * (n_points_fine)^dim n_points_coarse?

No that would be still guarded. The only difference would be that data is copied into a buffer in a vectorized form.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 1]
```diff
@@ -2308,6 +2428,38 @@ namespace MGTransferGlobalCoarseningTools
...
+          {
+            const unsigned int shift = 9 * degree_to_3[k] + 3 * degree_to_3[j];
+            data[0] *= weights[shift];
+            // loop bound as int avoids compiler warnings in case loop_length
+            // == 1 (polynomial degree 0)
+            for (int i = 1; i < loop_length - 1; ++i)
+              data[i] *= weights[shift + 1];
+            data[loop_length - 1] *= weights[shift + 2];
+            data += loop_length;
+          }
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:2457]: Is this the same code as in the other MG transfer function https://github.com/dealii/dealii/blob/cfa11cfe8cbc78f9a3b15bdb435fd2297d4b871d/source/multigrid/mg_transfer_matrix_free.cc#L362-L388

I suggest to move the code to a common location and only implement it once. Since this is a regular tensor product variant, it might fit in `include/deal.II/matrix_free/tensor_product_kernels.h`.

**Review by kronbichler — COMMENTED** [review]
Looks good to me, apart from a few comments on the code. I agree we should remove the specialization as it does not contribute to faster execution (unless everything sits in L2 cache).

**Comment by peterrum** [comment]
@kronbichler I have made the changes!

File: include/deal.II/matrix_free/tensor_product_kernels.h [hunk 2]
```diff
@@ -2586,6 +2586,36 @@ namespace internal
   }
 
 
+  template <int dim, typename Number>
+  inline void
+  weight_dofs_on_child(const VectorizedArray<Number> *weights,
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tensor_product_kernels.h:2591]: Can we rename this into something more generic, e.g. `weight_fe_q_dofs_by_entity`? Also, I wonder if you would mind injecting the dimension `loop_length` via a template argument for the case this is supported, in order to reduce loop overhead?

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/matrix_free/tensor_product_kernels.h (hunk 2 above)
Comment (peterrum) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2591]: > Also, I wonder if you would mind injecting the dimension loop_length via a template argument for the case this is supported, in order to reduce loop overhead?

I would have thought that inline is enough. But I can do the change.

Comment (kronbichler) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2591]: I am not sure, but my experience is that the compiler feels tempted to not inline such functions (with more than >20 obvious instructions) because it sees it fit for multiple template arguments, reducing the machine code size a bit. Now I won't blame any compiler for doing so (not sure if this is the case here, though), because it is impossible for the compiler to know that different degrees will rarely be executed in close temporal proximity, making the code size reduction effective at all.

Comment (peterrum) [reply include/deal.II/matrix_free/tensor_product_kernels.h]: Done!
